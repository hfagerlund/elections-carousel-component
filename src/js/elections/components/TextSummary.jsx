'use strict';
import React from 'react';
import PropTypes from 'prop-types';

import css from './textsummary.scss';

export default class TextSummary extends React.Component {
  _roundToTwoDec(num) {
    /*
     * source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round
     */
    return +(Math.round(num + 'e+2') + 'e-2');
  }

  _getTotalVotes(riding) {
    const totalVotes = riding
      ? riding.map(results => results.votes).reduce((a, b) => a + b, 0)
      : null;
    return totalVotes;
  }

  getCandidates(riding) {
    const listCandidates = riding
      ? riding.map(results => (
          <p
            key={results.name}
            className={`${css['candidate']}${results.isElected ? ' candidateWinner' : ''}`}
          >
            <span
              className={`label cap candidate_${results.partyCode.toLowerCase()}${
                results.isElected ? ' labelWinner' : ''
              }`}
            >
              {' '}
              {results.partyCode}{' '}
            </span>

            <span className={`${css['result']}${results.isElected ? ' summaryWinner' : ''}`}>
              {results.name} {results.isElected ? 'was elected with' : 'received'} {results.votes}{' '}
              votes ({this._roundToTwoDec(results.votes / this._getTotalVotes(riding) * 100)}%)
            </span>
          </p>
        ))
      : null;
    return listCandidates;
  }

  render() {
    return <div className={css.ridingSummary}>{this.getCandidates(this.props.riding.results)}</div>;
  }
}

TextSummary.propTypes = {
  riding: PropTypes.object.isRequired
};
