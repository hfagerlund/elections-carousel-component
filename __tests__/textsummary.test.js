import TextSummary from '../src/js/elections/components/TextSummary';
import React from 'react';

import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

test('TextSummary component renders a riding summary', () => {
  const tree = shallow(<TextSummary riding={{}} />);
  expect(tree).toMatchSnapshot();
});

test('TextSummary component matches its snapshot with riding data', () => {
  const ridings = {
    id: 1,
    name: 'Annapolis',
    num: 1,
    pollsReported: 57,
    pollsTotal: 57,
    results: [
      {
        name: 'Henry Spurr',
        partyId: 13,
        votes: 834,
        isElected: false,
        partyCode: 'NDP'
      },
      {
        name: 'Stephen McNeil',
        partyId: 12,
        votes: 7709,
        isElected: true,
        partyCode: 'LIB'
      },
      {
        name: 'Ginny Hurlock',
        partyId: 14,
        votes: 1390,
        isElected: false,
        partyCode: 'PC'
      },
      {
        name: 'Ron Neufeld',
        partyId: 11,
        votes: 227,
        isElected: false,
        partyCode: 'GRN'
      }
    ]
  };
  const tree = renderer.create(<TextSummary riding={ridings} />);
  expect(tree).toMatchSnapshot();
});
