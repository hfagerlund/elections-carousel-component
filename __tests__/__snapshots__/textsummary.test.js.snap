// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextSummary component matches its snapshot with riding data 1`] = `
<div
  className="ridingSummary"
>
  <p
    className="candidate"
  >
    <span
      className="label cap candidate_ndp"
    >
       
      NDP
       
    </span>
    <span
      className="result"
    >
      Henry Spurr
       
      received
       
      834
       
      votes (
      8.21
      %)
    </span>
  </p>
  <p
    className="candidate candidateWinner"
  >
    <span
      className="label cap candidate_lib labelWinner"
    >
       
      LIB
       
    </span>
    <span
      className="result summaryWinner"
    >
      Stephen McNeil
       
      was elected with
       
      7709
       
      votes (
      75.88
      %)
    </span>
  </p>
  <p
    className="candidate"
  >
    <span
      className="label cap candidate_pc"
    >
       
      PC
       
    </span>
    <span
      className="result"
    >
      Ginny Hurlock
       
      received
       
      1390
       
      votes (
      13.68
      %)
    </span>
  </p>
  <p
    className="candidate"
  >
    <span
      className="label cap candidate_grn"
    >
       
      GRN
       
    </span>
    <span
      className="result"
    >
      Ron Neufeld
       
      received
       
      227
       
      votes (
      2.23
      %)
    </span>
  </p>
</div>
`;

exports[`TextSummary component renders a riding summary 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextSummary
    riding={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "className": "ridingSummary",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "className": "ridingSummary",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
